dbms/src/Functions/FunctionsConsistentHashing.h

```diff
@@ -0,0 +1,173 @@
+    }
+};
+
+
+template <typename Impl>
+class FunctionConsistentHashImpl : public IFunction
+{
+public:
+
+    static constexpr auto name = Impl::name;
+    using ResultType = typename Impl::ResultType;
+    using BucketsType = typename Impl::BucketsCountType;
+
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionConsistentHashImpl<Impl>>(); };
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (!arguments[0]->isInteger())
+            throw Exception("Illegal type " + arguments[0]->getName() + " of the first argument of function " + getName(),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        if (!arguments[1]->isInteger())
+            throw Exception("Illegal type " + arguments[1]->getName() + " of the second argument of function " + getName(),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        return std::make_shared<DataTypeNumber<ResultType>>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) override
+    {
+        auto buckets_col = block.getByPosition(arguments[1]).column.get();
+        if (!buckets_col->isColumnConst())
+            throw Exception("The second argument of function " + getName() + " (number of buckets) must be constant", ErrorCodes::BAD_ARGUMENTS);
+
+        constexpr UInt64 max_buckets = static_cast<UInt64>(std::numeric_limits<BucketsType>::max());
+        UInt64 num_buckets;
+
+        auto check_range = [&] (auto buckets)
+        {
+            if (buckets <= 0)
+                throw Exception("The second argument of function " + getName() + " (number of buckets) must be positive number",
+                                ErrorCodes::BAD_ARGUMENTS);
+
+            if (static_cast<UInt64>(buckets) > max_buckets)
+                throw Exception("The value of the second argument of function " + getName() + " (number of buckets) is not fit to " +
+                            DataTypeNumber<BucketsType>().getName(), ErrorCodes::BAD_ARGUMENTS);
+
+            num_buckets = static_cast<UInt64>(buckets);
+        };
+
+        Field buckets_field = (*buckets_col)[0];
+        if (buckets_field.getType() == Field::Types::Int64)
+            check_range(buckets_field.safeGet<Int64>());
+        else if (buckets_field.getType() == Field::Types::UInt64)
+            check_range(buckets_field.safeGet<UInt64>());
+        else
+            throw Exception("Illegal type " + String(buckets_field.getTypeName()) + " of the second argument of function " + getName(),
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+
+        const auto & hash_col_source = block.getByPosition(arguments[0]).column;
+        ColumnPtr hash_col = (hash_col_source->isColumnConst()) ? hash_col_source->convertToFullColumnIfConst() : hash_col_source;
```

Comment:
No need for this. Look at `useDefaultImplementationForConstants` and `getArgumentsThatAreAlwaysConstant`.

Reply:
(The function should be constexpr for constant arguments.)